File: pkg/dart2cpp/sample/dart/test_simple_functions.dart
// 简单函数测试

// 1. 普通函数定义
int add(int a, int b) {
  return a + b;
}

int subtract(int a, int b) {
  return a - b;
}

// 2. Lambda 函数表达式
var multiply = (int a, int b) => a * b;

// 3. 函数作为参数传递
int calculate(int a, int b, Function operation) {
  return operation(a, b);
}

// 主函数
void main() {
  // 测试普通函数
  var result1 = calculate(10, 5, add);
  print('10 + 5 = $result1');
  
  // 测试 Lambda 函数
  var result2 = calculate(10, 5, multiply);
  print('10 * 5 = $result2');
}

File: pkg/dart2cpp/sample/output/test_simple_functions.cpp
#include "dart2cpp.h"

// 工具宏定义

Int add(Int a, Int b);
Int subtract(Int a, Int b);
Int calculate(Int a, Int b, Function operation);
Int add(Int a, Int b) {
  return (a + b);
}

Int subtract(Int a, Int b) {
  return (a - b);
}

Int calculate(Int a, Int b, Function operation) {
  return dart_cast<Int>(operation->apply({a, b}));
}

// ============================================================================
// 主函数
// ============================================================================

int main() {
  try {
    auto result1 = calculate(dart_int(10), dart_int(5), makeFunction(&add));
dart_print(dart_concat(dart_string("10 + 5 = "), result1));
auto result2 = calculate(dart_int(10), dart_int(5), multiply);
dart_print(dart_concat(dart_string("10 * 5 = "), result2));
    return 0;
  } catch (const std::exception& e) {
    std::cerr << "Error: " << e.what() << std::endl;
    return 1;
  }
}
